qiskit/providers/aer/backends/aer_compiler.py

```diff
@@ -29,37 +29,58 @@ class AerCompiler:
 
         Args:
-            circuits (QuantumCircuit or list): The QuantumCircuit (or list
-                of QuantumCircuit objects) to be compiled
-            basis_gates (list): basis gates to decompose sub-circuits(default: None).
+            circuits (QuantumCircuit or list): The QuantumCircuits to be compiled
+            basis_gates (list): basis gates to decompose sub-circuits
+                                (default: None).
+            optypes (list): list of instruction type sets for each circuit
+                            (default: None).
 
         Returns:
             QuantumCircuit or list: QuantumCircuit (or list
                 of QuantumCircuit objects) without control-flow instructions
         """
-        if basis_gates:
-            basis_gates = basis_gates + ['mark', 'jump']
+        if isinstance(circuits, QuantumCircuit):
+            circuits = [circuits]
         if isinstance(circuits, list):
-            return [transpile(self._inline_circuit(circ, None, None),
-                              basis_gates=basis_gates) if self._is_dynamic(circ)
-                    else circ for circ in circuits]
-        else:
-            return (transpile(self._inline_circuit(circuits, None, None),
-                              basis_gates=basis_gates) if self._is_dynamic(circuits)
-                    else circuits)
-
-    def _is_dynamic(self, circuit):
-        """check whether a circuit contains control-flow instructions
-        """
+            basis_gates = basis_gates + ['mark', 'jump']
+            compiled_circuits = []
+            for idx, circuit in enumerate(circuits):
+                optype = optypes[idx] if optypes else None
+                if isinstance(circuit, QuantumCircuit) and self._is_dynamic(
+                        circuit, optype):
+                    compiled_circ = transpile(
+                        self._inline_circuit(circuit, None, None),
+                        basis_gates=basis_gates
+                    )
+                    compiled_circuits.append(compiled_circ)
+                else:
+                    compiled_circuits.append(circuit)
+            return compiled_circuits
+        return circuits
+
+    @staticmethod
+    def _is_dynamic(circuit, optype=None):
+        """check whether a circuit contains control-flow instructions"""
         if not isinstance(circuit, QuantumCircuit):
             return False
+
+        controlflow_types = (
+            WhileLoopOp, ForLoopOp, IfElseOp, BreakLoopOp, ContinueLoopOp
+        )
+
+        # Check via optypes
+        if isinstance(optype, set) and optype.intersection(controlflow_types):
+            return True
+
+        # Check via iteration
         for inst, _, _ in circuit.data:
-            if isinstance(inst, (WhileLoopOp, ForLoopOp, IfElseOp, BreakLoopOp, ContinueLoopOp)):
+            if isinstance(inst, controlflow_types):
                 return True
+
         return False
```

Comment:
I think you want an early `False` return here if `optype` is a set, but there's no intersection.  Otherwise you're looping through the data even though you know it should fail.

Reply:
In this case i think returning True is the early case since its basically an `if any` check for a circuit containing any of hte control flow instructions.

Reply:
No, I agree with that point, the trick is that `if isinstance(optype, set) and not optype.intersection(controlflow_types)` then you know that the the iterative loop will return `False`.  If `optype` is not `None`, then you should never need to iterate through the instructions.

Reply:
Ah i see